edweiss-app/components/courses/AssignmentComponent.tsx

```diff
@@ -0,0 +1,267 @@
+    dateIcon: 'calendar',
+    timeIcon: 'alarm',
+    finishIcon: 'checkmark-circle',
+    deleteIcon: 'trash',
+};
+
+// Tests Tags
+export const testIDs: { [key: string]: string } = {
+    addAssignmentTitle: 'add-assignment-title',
+    addAssignmentDescription: 'add-assignment-description',
+    editAssignmentTitle: 'edit-assignment-title',
+    editAssignmentDescription: 'edit-assignment-description',
+    scrollView: 'scroll-view',
+    nameAndTypeView: 'name-and-type-view',
+    nameInput: 'name-input',
+    typeInput: 'type-input',
+    dueDateView: 'due-date-view',
+    dateInput: 'date-input',
+    dateTitle: 'date-title',
+    dateTouchableOpacity: 'date-touchable-opacity',
+    dateIcon: 'date-icon',
+    dateText: 'date-text',
+    timeInput: 'time-input',
+    timeTitle: 'time-title',
+    timeTouchableOpacity: 'time-touchable-opacity',
+    timeIcon: 'time-icon',
+    timeText: 'time-text',
+    datePicker: "dateTimePicker1",
+    timePicker: "dateTimePicker2",
+    finishViews: 'finish-views',
+    submitTouchableOpacity: 'submit-touchable-opacity',
+    submitView: 'submit-view',
+    submitIcon: 'submit-icon',
+    submitText: 'submit-text',
+    deleteTouchableOpacity: 'delete-touchable-opacity',
+    deleteView: 'delete-view',
+    deleteText: 'delete-text',
+    deleteIcon: 'delete-icon',
+};
+
+type AssignmentMode = 'edit' | 'add';
+
+interface EditAssignmentProps {
+    mode: AssignmentMode;
+    onSubmit: (assignment: Assignment, assignmentID?: AssignmentID) => void;
+    onDelete?: (assignmentID: AssignmentID) => void;
+    assignment?: { id: string, data: Assignment };
+}
+
+
+/**
+ * AssignmentComponent Component
+ * 
+ * This component is responsible for adding or editing an assignment in the course page.
+ * 
+ * @param mode - The mode of the assignment component.
+ * @param onSubmit - The function to call when the user submits the assignment.
+ * @param onDelete - The function to call when the user deletes the assignment.
+ * @param assignment - The assignment data to be edited.
+ * 
+ * @returns JSX.Element - The rendered component for the assignment editing.
+ */
+const AssignmentComponent: ReactComponent<EditAssignmentProps> = ({ mode, onSubmit, onDelete, assignment }) => {
+
+    const [name, setName] = React.useState<string>(assignment && mode == 'edit' ? assignment.data.name : "");
+    const [type, setType] = React.useState<AssignmentType>(assignment && mode == 'edit' ? assignment.data.type : 'quiz');
+    const [dueDate, setDueDate] = React.useState<Date>(assignment && mode == 'edit' ? Time.toDate(assignment.data.dueDate) : new Date());
+    const [nameChanged, setNameChanged] = useState<boolean>(false);
+    const [dateChanged, setDateChanged] = useState<boolean>(false);
+    const [timeChanged, setTimeChanged] = useState<boolean>(false);
+    const [showPickerDate, setShowPickerDate] = useState<boolean>(false);
+    const [showPickerTime, setShowPickerTime] = useState<boolean>(false);
```

Comment:
This is a great way to handle the codebase using use states. The things is that semantically the use states are not independant,. What I mean by this is context indenpant. They are all related to the `AssignementComponent` state.
In the futur, I would propose you this approach in order to pack things together and enhance code maintainability.

Refactor related states into a single `useReducer` or an object in `useState`. Here's what I mean:

```tsx
const [state, setState] = useState({
    name: assignment?.data.name || "",
    type: assignment?.data.type || "quiz",
    dueDate: assignment?.data.dueDate || new Date(),
    nameChanged: false,
    dateChanged: false,
    timeChanged: false,
    showPickerDate: false,
    showPickerTime: false,
});

const updateState = (key: string, value: any) => {
    setState((prev) => ({ ...prev, [key]: value }));
};

```

and you can use it this way to update the fields
```tsx

     updateState('name', 'new name');

```

Reply:
I slightly modified your version and integrated it, you can take a look if you want to. Thank you!

Reply:
Hello,
This is bad practice as each `setState` would require to reconstruct the object.
This does not provide any advantage over @FlorianDinant 's method, as `updateState('name', 'new name')` has no advantage over `setName('new name')`
Also, this version does not provide any Intellisense or type safety.
`updateState("namr", 2)` is a valid call, this is obviously bad.
Please do not implement this.